tests/fail/tree_borrows/read_retag_no_race.rs

```diff
@@ -0,0 +1,96 @@
+//@compile-flags: -Zmiri-tree-borrows -Zmiri-preemption-rate=0
+use std::ptr::addr_of_mut;
+use std::sync::{Arc, Barrier};
+use std::thread;
+
+#[derive(Copy, Clone)]
+struct SendPtr(*mut u8);
+
+unsafe impl Send for SendPtr {}
+
+// This test features the problematic pattern
+//
+// read x     || retag y (&mut, protect)
+//        -- sync --
+//            || write y
+//
+// In which
+// - one interleaving (`1:read; 2:retag; 2:write`) does not have UB if retags
+//   count only as reads for the data race model,
+// - the other interleaving (`2:retag; 1:read; 2:write`) has UB (`noalias` violation).
+//
+// The interleaving executed here is the one that has UB,
+// i.e.
+//      retag x
+//      read y
+//      write x
+//
+// This is a clear violation of the `noalias` requirements on `x`,
+// but we are ensuring that the UB is detected by the aliasing model
+// and *not* by the data race model.
+// Although there is race-like behavior going on in this pattern
+// (where some but not all interleavings contain UB), making this an actual
+// data race has the confusing requirement of one single access being treated
+// as being of different `AccessKind`s by different parts of Miri.
+
+// The other interleaving is featured in tests/pass/tree_borrows/read_retag_no_race.rs
```

Comment:
Can't you have both in the same test (in a single function with a parameter that says which one to use), and then in the test do
```
fn main() {
  run_test_without_race();
  run_test_with_race();
}
```
and then the backtrace will show that only the 2nd test fails, which means the first one passed?

Reply:
In fact I don't think I understand the point of this test... "not a write for the data race model" leads to less UB so I only expected a pass test, not a fail test?

Reply:
With write-on-retag for the data race model, both interleavings were UB. Not doing writes for the data race model has made exactly one of those two interleavings not contain UB anymore, so I was checking that.

As I mentioned, the fail test is also a subpattern of `tests/fail/tree_borrows/spurious_read.rs`: we already have a test that contains a `retag x; read y; write x` and checks that it is UB on the write (though it is only identical up to an α-renaming of `x`,`y` into `y`,`x` which is confusing and which I will fix), so I could have `tests/pass/tree_borrows/read_retag_no_race.rs` refer to `tests/fail/tree_borrows/spurious_read.rs` as its companion interleaving and remove `tests/fail/tree_borrows/read_retag_no_race.rs`.

There is however a bigger problem with `tests/pass/tree_borrows/read_retag_no_race.rs` test, which is that I've probably inserted too much synchonization that even if we re-enabled the data-race-write-on-retag it wouldn't be a race. I'll need to rework the entire test anyway.